include/aws/common/file.h

```diff
@@ -4,19 +4,108 @@
  * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
  * SPDX-License-Identifier: Apache-2.0.
  */
+#include <aws/common/byte_buf.h>
 #include <aws/common/common.h>
+#include <aws/common/linked_list.h>
+#include <aws/common/ref_count.h>
 
-AWS_EXTERN_C_BEGIN
+#ifdef _WIN32
+#    define AWS_PATH_DELIM '\\'
+#    define AWS_PATH_DELIM_STR "\\"
+#else
+#    define AWS_PATH_DELIM '/'
+#    define AWS_PATH_DELIM_STR "/"
+#endif
+
+struct aws_string;
+
+enum aws_file_type {
+    AWS_FILE_TYPE_FILE = 1,
+    AWS_FILE_TYPE_SYM_LINK = 2,
+    AWS_FILE_TYPE_DIRECTORY = 4,
+};
+
+struct aws_directory_entry {
+    /**
+     * Absolute path to the entry from the current process root.
+     */
+    struct aws_byte_cursor path;
+    /**
+     * Path to the entry relative to the current working directory.
+     */
+    struct aws_byte_cursor relative_path;
```

Comment:
if we're going to require users to pass `aws_string*` or `const char *` to our file APIs, we may as well make `path` and `relative_path` either `aws_string *` or `const char *` too

Reply:
don't need to though. I can guarantee they're properly constructed. This does nothing except for increase allocaton overhead for no good reason.

Reply:
the api and type system will already force users to pass an aws_string to turn around and call the other apis. I don't know, I see the argument on ergonomics, but I don't think the ergonomics make practical sense here unless we go rewrite the aws_string implementation.